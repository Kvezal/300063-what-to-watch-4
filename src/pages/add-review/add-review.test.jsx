import React from "react";
import render from "react-test-renderer";
import Enzyme, {shallow} from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import AddReview from "./add-review";
import {MemoryRouter} from "react-router-dom";


const film = {
  id: 1,
  name: `name 1`,
  genre: `Crime`,
  runTime: `3h 49m`,
  releaseDate: 1984,
  description: `description 1`,
  director: `director 1`,
  rating: {
    score: 9.9,
    count: 276395,
    assessment: `Very good`,
  },
  source: {
    video: `video1`,
    previewVideo: `previewVideo1`,
  },
  picture: {
    poster: `poster1`,
    preview: `preview1`,
    cover: `cover1`,
    backgroundColor: `color1`,
  },
  starring: [
    `Robert De Niro`,
    `James Woods`,
    `Elizabeth McGovern`
  ],
};

Enzyme.configure({
  adapter: new Adapter(),
});

describe(`AddReviewPage`, () => {
  test(`should render component`, () => {
    const tree = render.create(
        <MemoryRouter>
          <AddReview
            avatar="test-avatar"
            isAuthorized={true}
            film={film}
            onSubmitForm={() => {}}
          />
        </MemoryRouter>
    )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test(`should create component`, () => {
    const addReviewComponent = shallow(
        <AddReview
          avatar="test-avatar"
          isAuthorized={true}
          film={film}
          onSubmitForm={() => {}}
        />
    );
    const addReview = addReviewComponent.find(`.add-review`);
    expect(addReview).toHaveLength(1);
  });

  test(`should disabled button`, () => {
    const addReviewComponent = shallow(
      <AddReview
        avatar="test-avatar"
        isAuthorized={true}
        film={film}
        onSubmitForm={() => {}}
      />
    );
    const buttonRef = addReviewComponent.instance()._buttonRef = {
      current: {
        disabled: false,
      }
    };
    addReviewComponent.instance()._ratingRef.current = {
      value: `1`,
    };
    addReviewComponent.instance()._commentRef.current = {
      value: new Array(49).fill(0).join(``),
    };
    addReviewComponent.find(`.add-review__textarea`).simulate(`change`);
    expect(buttonRef.current.disabled).toBeTruthy();

    addReviewComponent.instance()._commentRef.current = {
      value: new Array(50).fill(0).join(``),
    };
    addReviewComponent.find(`.add-review__textarea`).simulate(`change`);
    expect(buttonRef.current.disabled).toBeFalsy();

    addReviewComponent.instance()._commentRef.current = {
      value: new Array(400).fill(0).join(``),
    };
    addReviewComponent.find(`.add-review__textarea`).simulate(`change`);
    expect(buttonRef.current.disabled).toBeFalsy();

    addReviewComponent.instance()._commentRef.current = {
      value: new Array(401).fill(0).join(``),
    };
    addReviewComponent.find(`.add-review__textarea`).simulate(`change`);
    expect(buttonRef.current.disabled).toBeTruthy();
  });

  test(`should submit form with correct params`, () => {
    const onSubmitForm = jest.fn();
    const rating = 5;
    const comment = new Array(100).fill(1).join(``);
    const addReviewComponent = shallow(
      <AddReview
        avatar="test-avatar"
        isAuthorized={true}
        film={film}
        onSubmitForm={onSubmitForm}
      />
    );
    addReviewComponent.instance()._ratingRef.current = {value: rating};
    addReviewComponent.instance()._commentRef.current = {value: comment};
    addReviewComponent.find(`form`).simulate(`submit`, {preventDefault: () => {}});
    expect(onSubmitForm).toBeCalledTimes(1);
    expect(onSubmitForm.mock.calls[0][0]).toEqual({rating, comment});
  });
});
