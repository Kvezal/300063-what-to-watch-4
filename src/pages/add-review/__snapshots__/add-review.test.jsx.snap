// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddReviewPage should render component 1`] = `
<section
  className="movie-card movie-card--full"
  style={
    Object {
      "background": "color1",
    }
  }
>
  <div
    className="movie-card__header"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="name 1"
        src="cover1"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <nav
        className="breadcrumbs"
      >
        <ul
          className="breadcrumbs__list"
        >
          <li
            className="breadcrumbs__item"
          >
            <a
              aria-current={null}
              className="breadcrumbs__link"
              href="/films/1"
              onClick={[Function]}
            >
              name 1
            </a>
          </li>
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
            >
              Add review
            </a>
          </li>
        </ul>
      </nav>
      <div
        className="user-block"
      >
        <div
          className="user-block__avatar"
        >
          <img
            alt="User avatar"
            height="63"
            src="test-avatar"
            width="63"
          />
        </div>
      </div>
    </header>
    <div
      className="movie-card__poster movie-card__poster--small"
    >
      <img
        alt="The Grand Budapest Hotel poster"
        height="327"
        src="poster1"
        width="218"
      />
    </div>
  </div>
  <div
    className="add-review"
  >
    <form
      action="#"
      className="add-review__form"
      onSubmit={[Function]}
    >
      <div
        className="rating"
      >
        <div
          className="rating__stars"
        >
          <input
            checked={false}
            className="rating__input"
            id="star-1"
            name="rating"
            onChange={[Function]}
            type="radio"
            value="1"
          />
          <label
            className="rating__label"
            htmlFor="star-1"
          />
          <input
            checked={false}
            className="rating__input"
            id="star-2"
            name="rating"
            onChange={[Function]}
            type="radio"
            value="2"
          />
          <label
            className="rating__label"
            htmlFor="star-2"
          />
          <input
            checked={true}
            className="rating__input"
            id="star-3"
            name="rating"
            onChange={[Function]}
            type="radio"
            value="3"
          />
          <label
            className="rating__label"
            htmlFor="star-3"
          />
          <input
            checked={false}
            className="rating__input"
            id="star-4"
            name="rating"
            onChange={[Function]}
            type="radio"
            value="4"
          />
          <label
            className="rating__label"
            htmlFor="star-4"
          />
          <input
            checked={false}
            className="rating__input"
            id="star-5"
            name="rating"
            onChange={[Function]}
            type="radio"
            value="5"
          />
          <label
            className="rating__label"
            htmlFor="star-5"
          />
        </div>
      </div>
      <div
        className="add-review__text"
        style={
          Object {
            "background": "rgba(20, 20, 20, 1)",
          }
        }
      >
        <textarea
          className="add-review__textarea"
          id="review-text"
          name="review-text"
          onChange={[Function]}
          placeholder="Review text"
        />
        <div
          className="add-review__submit"
        >
          <button
            className="add-review__btn"
            disabled={true}
            style={
              Object {
                "color": "rgba(255, 255, 255, 1)",
              }
            }
            type="submit"
          >
            Post
          </button>
        </div>
      </div>
    </form>
  </div>
</section>
`;
